Rename CurvedSW execs, move is_analytic_*, other analytic solution updates

nilsdeppe opened the pull request:
## Proposed changes

First steps to being able to much more easily observe things when using DG or DG-subcell. Ultimately will also be able to remove `ObserveErrorNorms` (few PRs away, still).

Most of the diff is moving code, which is done in its own commits.

- Rename the CurvedSW executables to be consistent with other systems
- Move `is_analytic_data/solution` out of the evolution directory
- Remove `InitializeAnalyticSolution` from elliptic code. Now always initializes as an optional
- Switch analytic solution and error tags to use `std::optional<Variables>`
- Add some helpers for working with `std::optional` like `has_value()` and `value()` free functions. These will be needed in a later PR where I refactor ObserveFields and ObserveNorms
- Switch GH executable to use the `ObservationBox`. This reduces memory usage and will generally make interplay with subcell easier.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Move is_analytic_data/solution to PointwiseFunctions

This was in evolution before, which seems like the wrong place.

nilsdeppe committed:
Switch AnalyticSolutions and error tags to optionals

This is necessary for controlling the initial data from the input file. The tags
evaluate the solution if it exists, otherwise results are std::nullopt.

nilsdeppe committed:
Remove InitializeAnalyticSolution from elliptic code

This simplifies the code for handling analytic things quite a bit since now it's
always a std::optional. I expect there to be more places that would benefit from
the guarantee that the solution variables are always std::optionals.

nilsdeppe committed:
Add std::optional helpers

nilsdeppe committed:
Switch GH executable to using the ObservationBox more

Reduces persistent memory usage

nilsdeppe committed:
Rename CurvedScalarWave executables to be consistent

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/AnalyticSolution.hpp:
```diff
@@ -3,7 +3,24 @@
+/// \ingroup AnalyticSolutionsGroup
+/// \brief Check if `T` is an analytic solution
+template <typename T>
+using is_analytic_solution =
```
We will probably delete this file once we have moved to factory-constructed initial data, correct?

nilsdeppe replied:
No, because you still need to know if the derived class is called with or without a `time` argument

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Tags.hpp:
```diff
@@ -11,6 +11,7 @@
 #include "Options/Options.hpp"
 #include "Parallel/Serialize.hpp"
 #include "PointwiseFunctions/AnalyticData/Tags.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/AnalyticSolution.hpp"
```
Is this actually needed?

nilsdeppe replied:
I think I can move it to the next commit

nilsvu commented on src/Evolution/ComputeTags.hpp:
```diff
@@ -11,81 +15,50 @@
+    } else {
+      (void)analytic_solution_computer;
+      (void)inertial_coords;
+      (void)time;
```
(optional) I prefer `[[maybe_unused]]` these days

nilsdeppe replied:
I don't like it in these cases because I prefer to explicitly acknowledge which arguments I'm not using in the branch

nilsvu commented on src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp:
```diff
@@ -313,12 +313,11 @@ struct EvolutionMetavars {
+      tmpl::conditional_t<is_analytic_solution_v<initial_data>,
+                          Initialization::Actions::AddComputeTags<tmpl::list<
+                              evolution::Tags::AnalyticSolutionsCompute<
+                                  Dim, analytic_solution_fields>>>,
```
Is the plan to move this compute tag to the ObservationBox? Maybe add a separate commit to this PR for the executables that support this already?

nilsdeppe replied:
Yes. I don't see a reason to move that code to this PR vs keeping it in the followup where it's actually needed. That's a lot of work for a net-zero effect as far as I can tell.

nilsvu replied:
Sure, follow-up is fine. Just checking if I understand the plan correctly.

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Tags.hpp:
```diff
@@ -80,18 +73,231 @@ struct Error : db::PrefixTag, db::SimpleTag {
-/// The analytic solutions for all `FieldTags`
+namespace detail {
+template <typename Tag>
+struct AnalyticImpl : db::PrefixTag, db::SimpleTag {
```
You use this tag quite a lot outside this file, so `detail` and `Impl` seems inappropriate. How about this tag remains `::Tags::Analytic`, so you can drop the renames in other files? Then the `std::optional` tag above could become `::Tags::AnalyticOptional`.

nilsdeppe replied:
It's only used a lot right now because the observers still compute errors locally. Once that's no longer the case it's only used inside this file and in `src/Elliptic/BoundaryConditions/AnalyticSolution.hpp` (I don't count tests because things need to be tested). I find the naming you suggested terribly confusing, because then you get things from the `DataBox` using `Tags::AnalyticOptional` while I would expect to use `Tags::Analytic`.

nilsvu replied:
- Don't worry about `Elliptic/BoundaryConditions/AnalyticSolution.hpp`. I'll refactor it to store its own factory-created solution object.
- I'm worried about placing a Variables in a DataBox tag, which can only be used with private/detail/impl tags to access the constituents. I guess it's fine if it lives only in the ObservationBox, and accessing the constituents happens only ever through the subitems. I think it would be good to document that, if that is your intention.

nilsdeppe replied:
- Sounds good :)
- Ah yes, good point. I'll document that unlike traditional Variables, this (and error) use different subitems than might be expected, why that is, and how the type traits `is_analytic_solutions` plays into all of that. I'll have the type traits point to the simple tags for the documentation.

nilsvu commented on src/Evolution/ComputeTags.hpp:
```diff
@@ -11,81 +15,50 @@
+/// \cond
+namespace Tags {
+template <size_t Dim, typename FieldTagsList>
+struct is_analytic_solutions<
```
What do you need this for? Please add a comment here, briefly mentioning why you need to specialize the template here.

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Tags.hpp:
```diff
@@ -80,18 +73,231 @@ struct Error : db::PrefixTag, db::SimpleTag {
+};
+
+template <typename Tag>
+struct is_analytic_solutions : std::false_type {};
```
Please add a comment what you need this for. I suggest to name this `is_analytic_solutions_tag`.

nilsdeppe replied:
it's already in the `Tags` namespace, so I don't see a reason to add more qualifiers to the name.

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Tags.hpp:
```diff
@@ -80,18 +73,231 @@ struct Error : db::PrefixTag, db::SimpleTag {
+  using return_type = typename base::type;
+  // using argument_tags =
+  //     tmpl::list<AnalyticSolutions<field_tags>,
+  //     ::Tags::Variables<field_tags>>;
```
Why did you go with the other implementation? With this one, we could drop the subitems logic altogether, and also remove the single-tensor `std::optional` tags, right?

nilsdeppe replied:
1. Share more code with the elliptic solver
2. Needed for GH+GRMHD because the vars for GH are in `variables_tag` while the vars for GRMHD are in `primitive_variables_tag`, so there isn't one single `Variables` that holds all the analytic tags.


I will remove the commented out code.

No, I don't think we should try to remove the single-tensor `std::optional` tags because someone will want to observe the analytic solution, not just the error.

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Tags.hpp:
```diff
@@ -80,18 +73,231 @@ struct Error : db::PrefixTag, db::SimpleTag {
+};
+
+template <typename Tag>
+struct is_errors : std::false_type {};
```
same comment as above for `is_analytic_solutions`

nilsvu reviewed (changes requested):
I stopped reviewing here for now with a couple of questions. Will review the remaining commits later.

kidder commented on src/Elliptic/Actions/InitializeAnalyticSolution.hpp:
```diff
@@ -33,10 +33,7 @@ namespace elliptic::Actions {
- * analytic solution is available, and `InitializeOptionalAnalyticSolution` if
- * that is a runtime decision, e.g. based on a choice in the input file. The
- * `::Tags::AnalyticSolutionsBase` tag can be retrieved from the DataBox in
+ * The `::Tags::AnalyticSolutionsBase` tag can be retrieved from the DataBox in
```
With the previous sentence deleted, I don't know what `in either case` means...

nilsdeppe commented:
I've rebased on develop to get the recent fixes. I also pushed fixups. Thanks for the reviews!

nilsvu commented on src/Evolution/ComputeTags.hpp:
```diff
@@ -11,81 +15,50 @@
-    *analytic_solution =
-        variables_from_tagged_tuple(analytic_solution_computer.variables(
-            inertial_coords, time, AnalyticFieldsTagList{}));
+    if constexpr (std::is_convertible_v<AnalyticSolution*,
```
Use `is_analytic_solution_v`?

nilsvu commented on src/Evolution/ComputeTags.hpp:
```diff
@@ -11,81 +15,50 @@
-struct AnalyticCompute : ::Tags::AnalyticSolutions<AnalyticFieldsTagList>,
-                         db::ComputeTag {
+template <size_t Dim, typename AnalyticFieldsTagList>
+struct AnalyticSolutionsCompute
```
Rename the file to `AnalyticSolutionsCompute.hpp`? (probably easiest in a separate commit)

nilsdeppe replied:
I'd like to avoid a rename to minimize subtle conflicts in the next series of PRs... We probably need to restructure how we do evolution tags more broadly anyway

nilsvu replied:
Ok 👍

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Tags.hpp:
```diff
@@ -79,18 +72,231 @@ struct Error : db::PrefixTag, db::SimpleTag {
+      EXPAND_PACK_LEFT_TO_RIGHT(compute_error(
+          make_not_null(
+              &get<::Tags::detail::ErrorImpl<FieldTags>>(errors->value())),
+          vars, analytic_vars));
```
(nitpick) I'd take `analytic_vars.value()` here already, so all arguments to the lambda are non-optional.

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Tags.hpp:
```diff
@@ -79,18 +72,231 @@ struct Error : db::PrefixTag, db::SimpleTag {
+namespace Tags {
+template <typename Tag, typename ParentTag>
+struct Subitem<Tag, ParentTag,
+               Requires<is_analytic_solutions<ParentTag>::value>>
```
I think you don't need the `is_analytic_solutions` struct with its specialization for the compute tag (which is outside this file) if you implement a `detail` function like this here instead:

```cpp
// Check if the argument is a `::Tags::AnalyticSolutions` template, or derived
// from it
template <typename FieldTags>
constexpr std::true_type is_analytic_solutions(
    ::Tags::AnalyticSolutions<FieldTags>&&) {
  return {};
}
constexpr std::false_type is_analytic_solutions(...) { return {}; }
template <typename Tag>
static constexpr bool is_analytic_solutions_v =
    decltype(is_analytic_solutions(std::declval<Tag>()))::value;
```

This is how `::Tags::Variables` subitems are implemented. It keeps the implementation within this file.

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Tags.hpp:
```diff
@@ -79,18 +72,231 @@ struct Error : db::PrefixTag, db::SimpleTag {
+  static void function(const gsl::not_null<typename base::type*> result,
+                       const typename parent_tag::type& vars) {
+    if (vars.has_value()) {
+      *result = get<typename detail::AnalyticImpl<typename Tag::tag>>(*vars);
```
do I understand correctly that each subitem holds a copy of the tensor from the Variables?

nilsdeppe replied:
Yes, I've changed this to setting the pointers in a fixup (I'll ping once pushed)

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Tags.hpp:
```diff
@@ -79,18 +72,231 @@ struct Error : db::PrefixTag, db::SimpleTag {
+  using type = db::wrap_tags_in<::Tags::Analytic, field_tags>;
+
+  template <typename Subtag,
+            typename LocalTag = ::Tags::AnalyticSolutions<field_tags>>
```
(nitpick) I think you can do `LocalTag = Tag`

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Tags.hpp:
```diff
@@ -79,18 +72,231 @@ struct Error : db::PrefixTag, db::SimpleTag {
+      const gsl::not_null<typename LocalTag::type*> parent_value,
+      const gsl::not_null<typename Subtag::type*> sub_value) {
+    if (parent_value->has_value()) {
+      auto& vars = get<::Tags::detail::AnalyticImpl<typename Subtag::tag>>(
```
I think it would be clearer to name `vars` -> `tensor` or something like that, since it's not a `Variables`. Also below `vars_it` -> `tensor_component` or similar.

nilsvu commented on tests/Unit/Evolution/Test_ComputeTags.cpp:
```diff
@@ -37,71 +37,52 @@ struct AnalyticSolution {
+            evolution::Tags::AnalyticSolutionsCompute<1, tmpl::list<FieldTag>>,
+            Tags::ErrorsCompute<tmpl::list<FieldTag>>>>(
+        inertial_coords, TestAnalyticData{}, current_time, vars);
+    const DataVector expected{2., 4., 6., 8.};
```
delete this line

nilsvu commented on tests/Unit/Evolution/Test_ComputeTags.cpp:
```diff
@@ -37,71 +37,52 @@ struct AnalyticSolution {
-      "Variables(Error(FieldTag))");
+  TestHelpers::db::test_compute_tag<
+      evolution::Tags::AnalyticSolutionsCompute<1, tmpl::list<FieldTag>>>(
+      "AnalyticSolutions");
```
Also add this for `ErrorsCompute`

nilsvu replied:
<s>ping</s> nvm, I didn't realize this is in a different file.

nilsvu commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Test_Tags.cpp:
```diff
@@ -28,9 +61,47 @@ SPECTRE_TEST_CASE("Unit.AnalyticSolutions.Tags", "[Unit][PointwiseFunctions]") {
+
+    const auto box = db::create<
+        db::AddSimpleTags<domain::Tags::Coordinates<1, Frame::Inertial>,
+                          ::Tags::AnalyticSolution<TestAnalyticSolution>,
```
In this particular test I think you don't need the analytic solution class. I think you can delete if from this file, since it's only used for the `AnalyticSolutionsCompute` tag that's tested in another file. Here you are just working with `Variables`.

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Tags.hpp:
```diff
@@ -110,37 +131,51 @@ struct ErrorImpl : db::PrefixTag, db::SimpleTag {
+/// \endcond
 
+/*!
+ * \brief Compute tag for compute the error from the `Tags::Analytic` of the
```
delete the second "compute"?

nilsvu reviewed (changes requested):
Reviewed all commits now. Looks good, thanks for the added docs. Please squash existing fixups. A few more comments, in particular I think you can make the `is_analytic_solutions` type trait local to the subitems implementation.

nilsvu commented:
Needs a rebase on #3780

nilsdeppe commented:
Just pushed a fixup and rebased on develop. Thanks for the review!

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Tags.hpp:
```diff
@@ -193,41 +195,40 @@ struct ErrorsCompute<tmpl::list<FieldTags...>>
-template <typename Tag>
-struct is_errors : std::false_type {};
+namespace detail {
+// Check if the argument is a `::Tags::AnalyticSolutions` template, or derived
```
AnalyticSolutions -> Errors

nilsvu reviewed (dismissed):
LGTM you can squash (including this)

nilsdeppe commented:
rebased on develop, squashed, squashed in removal of dox saying that type traits need to be specialized since that's no longer the case.

Thanks for the review!

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/Tags.hpp:
```diff
@@ -60,37 +63,299 @@ struct AnalyticSolution : AnalyticSolutionBase, db::SimpleTag {
+};
+
+/// Base tag for the analytic solution tensors. Retrieved values can be either
+/// `Variables` or `std::optional<Variables>`.
```
delete this sentence

nilsvu reviewed (dismissed):
LGTM, one thing to squash in

nilsvu commented:
@kidder please take another look.

kidder commented on src/Evolution/Executables/Burgers/EvolveBurgers.hpp:
```diff
@@ -129,11 +129,10 @@ struct EvolutionMetavars {
   static_assert(
-      evolution::is_analytic_data_v<initial_data> xor
-          evolution::is_analytic_solution_v<initial_data>,
+      is_analytic_data_v<initial_data> xor is_analytic_solution_v<initial_data>,
```
add new includes

kidder commented on src/Evolution/Executables/CurvedScalarWave/EvolveCurvedScalarWave.hpp:
```diff
@@ -125,12 +125,11 @@ template <size_t Dim, typename InitialData>
   static_assert(
-      evolution::is_analytic_data_v<InitialData> xor
-          evolution::is_analytic_solution_v<InitialData>,
+      is_analytic_data_v<InitialData> xor is_analytic_solution_v<InitialData>,
```
add new includes; remove old include

kidder replied:
ping remove line 40

kidder commented on src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp:
```diff
@@ -263,8 +263,7 @@ struct GhValenciaDivCleanTemplateBase<
   static_assert(
-      evolution::is_analytic_data_v<initial_data> xor
-          evolution::is_analytic_solution_v<initial_data>,
+      is_analytic_data_v<initial_data> xor is_analytic_solution_v<initial_data>,
```
add new includes

kidder commented on src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivClean.hpp:
```diff
@@ -191,15 +191,14 @@ struct EvolutionMetavars {
   static_assert(
-      evolution::is_analytic_data_v<initial_data> xor
-          evolution::is_analytic_solution_v<initial_data>,
+      is_analytic_data_v<initial_data> xor is_analytic_solution_v<initial_data>,
```
add new, remove old includes

kidder replied:
ping remove line 79

kidder commented on src/Evolution/Executables/NewtonianEuler/EvolveNewtonianEuler.hpp:
```diff
@@ -138,8 +138,7 @@ struct EvolutionMetavars {
   static_assert(
-      evolution::is_analytic_data_v<initial_data> xor
-          evolution::is_analytic_solution_v<initial_data>,
+      is_analytic_data_v<initial_data> xor is_analytic_solution_v<initial_data>,
```
add new includes

kidder commented on src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp:
```diff
@@ -114,8 +113,7 @@ struct EvolutionMetavars {
   static_assert(
-      evolution::is_analytic_data_v<initial_data> xor
-          evolution::is_analytic_solution_v<initial_data>,
+      is_analytic_data_v<initial_data> xor is_analytic_solution_v<initial_data>,
```
add new, remove old includes

kidder commented on src/Evolution/Executables/RelativisticEuler/Valencia/EvolveValencia.hpp:
```diff
@@ -115,8 +115,7 @@ struct EvolutionMetavars {
   static_assert(
-      evolution::is_analytic_data_v<initial_data> xor
-          evolution::is_analytic_solution_v<initial_data>,
+      is_analytic_data_v<initial_data> xor is_analytic_solution_v<initial_data>,
```
add new includes

kidder commented on src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp:
```diff
@@ -131,11 +131,10 @@ struct EvolutionMetavars {
   static_assert(
-      evolution::is_analytic_data_v<initial_data> xor
-          evolution::is_analytic_solution_v<initial_data>,
+      is_analytic_data_v<initial_data> xor is_analytic_solution_v<initial_data>,
```
add new includes

kidder commented on src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp:
```diff
@@ -115,8 +114,7 @@ struct EvolutionMetavars {
   static_assert(
-      evolution::is_analytic_data_v<InitialData> xor
-          evolution::is_analytic_solution_v<InitialData>,
+      is_analytic_data_v<InitialData> xor is_analytic_solution_v<InitialData>,
```
add new includes

kidder commented on tests/Unit/Evolution/Test_TypeTraits.cpp:
```diff
@@ -1,33 +0,0 @@
-              "Failed testing evolution::is_analytic_data_v");
-static_assert(evolution::is_analytic_data_v<SolutionDependentAnalyticData>,
-              "Failed testing evolution::is_analytic_data_v");
-static_assert(
```
keep this (and the similar one below) in one of the places you moved the others

kidder reviewed (changes requested):
you can squash immediately

nilsdeppe commented:
Rebased and squashed. Thanks for the reviews!

kidder reviewed (changes requested):
you can squash
